### PR TITLE
Add explicit tracking of Airbyte Protocol Version in ConnectorSpecification

### PR DESCRIPTION
**Pull request by gosusnp**
## What

Airbyte does not currently support transitions of the Airbyte Protocol evolution. As a result, we require both the platform and the connectors to always support the same version of the protocol and we lack the ability to clearly delineate which connectors work with which versions of the platform, and communicate this to our users. We want to enable progressive update of the platform and the connectors. 

… enter Airbyte Protocol Versioning

## How

Add an optional `protocol_version` attribute to the `ConnectorSpecification`.
This attribute should describe the Airbyte Protocol version supported by the Connector.

For backwards compatibility, if the `protocol_version` is not provided, it should be treated as `0.2.0`.

### CODE REVIEW

**Review by evantahler — APPROVED** [review]
I think there are also some places to change in the docs too, like:
* https://docs.airbyte.com/understanding-airbyte/airbyte-protocol#spec
* and other place we might show example SPEC Json outoup

**Comment by gosusnp** [comment]
@evantahler, our current structure is focused on a single version.
I was thinking that we could do that later given that the current change remains technically non-breaking.

Should we postpone the doc update to when we actually do that structure change?

**Comment by evantahler** [comment]
> Docs

I do think we need /something/.  Perhpas at the very least we add a CHANGELOG section to the bottom of https://docs.airbyte.com/understanding-airbyte/airbyte-protocol. It could be a small table like the connector changelogs.  
* Link to the PR that made the change
* date
* version number
* short description

**Comment by gosusnp** [comment]
@evantahler, I updated the yaml description in the doc.
I'd rather update the doc later, I think the quick addition of a changelog would probably add more confusion since there's already some form of version attached to AirbyteStateMessage in parts of the doc.

I propose to clean this as part of the better doc for versioning in https://github.com/airbytehq/airbyte/issues/16024

File: airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml [hunk 1]
```diff
@@ -427,6 +427,9 @@ definitions:
             type: string
           oauth_config_specification:
             "$ref": "#/definitions/OAuthConfigSpecification"
+      protocol_version:
+        description: "the Airbyte Protocol version supported by the connector"
```
Comment (sherifnada) [inline airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:431]:
```suggestion
        description: "the Airbyte Protocol version supported by the connector. Protocol versioning uses SemVer. "
```

File: airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py [hunk 2]
```diff
@@ -278,6 +278,7 @@ class Config:
         None,
         description="Additional and optional specification object to describe what an 'advanced' Auth flow would need to function.\n  - A connector should be able to fully function with the configuration as described by the ConnectorSpecification in a 'basic' mode.\n  - The 'advanced' mode provides easier UX for the user with UI improvements and automations. However, this requires further setup on the\n  server side by instance or workspace admins beforehand. The trade-off is that the user does not have to provide as many technical\n  inputs anymore and the auth process is faster and easier to complete.",
     )
+    protocol_version: Optional[str] = Field(None, description="the Airbyte Protocol version supported by the connector")
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py:281]: Could you add a line to the CDK's CHANGELOG under the `Unreleased` header which says this was added? It'll get released on the next CDK change

**Review by sherifnada — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py (hunk 2 above)
Comment (gosusnp) [reply airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py:281]: Updated, see [CHANGELOG.md](https://github.com/airbytehq/airbyte/pull/15340/files#diff-eaac011381105d313e921677aa92d6099b5e258b3b66dba8ee778d52b2127baf)

**Pull request merged**